### PR TITLE
Deprecate old instructiont API

### PR DESCRIPTION
**Pull request by kroening**
No change to code, documentation only.

- [X] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- n/a Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- n/a My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: jbmc/src/java_bytecode/remove_instanceof.cpp [hunk 1]
```diff
@@ -211,7 +211,7 @@ bool remove_instanceoft::lower_instanceof(
     // GOTO programs before the target instruction without inserting into the
     // wrong basic block.
     goto_program.insert_before_swap(target);
-    target->make_skip();
+    *target = goto_programt::make_skip();
```
Comment (tautschnig) [inline jbmc/src/java_bytecode/remove_instanceof.cpp]: Why not `turn_into_skip`?

File: src/goto-instrument/accelerate/disjunctive_polynomial_acceleration.cpp [hunk 2]
```diff
@@ -915,7 +914,7 @@ void disjunctive_polynomial_accelerationt::build_fixed()
     if(loop.find(t)==loop.end())
     {
       // This instruction isn't part of the loop...  Just remove it.
-      fixedt->make_skip();
+      *fixedt = goto_programt::make_skip(fixedt->source_location);
```
Comment (tautschnig) [inline src/goto-instrument/accelerate/disjunctive_polynomial_acceleration.cpp]: `turn_into_skip`

File: src/goto-instrument/accelerate/sat_path_enumerator.cpp [hunk 3]
```diff
@@ -269,7 +269,7 @@ void sat_path_enumeratort::build_fixed()
     if(loop.find(t)==loop.end())
     {
       // This instruction isn't part of the loop...  Just remove it.
-      fixedt->make_skip();
+      *fixedt = goto_programt::make_skip(fixedt->source_location);
```
Comment (tautschnig) [inline src/goto-instrument/accelerate/sat_path_enumerator.cpp]: `turn_into_skip`

File: src/goto-instrument/generate_function_bodies.cpp [hunk 4]
```diff
@@ -74,15 +74,13 @@ class assume_false_generate_function_bodiest : public generate_function_bodiest
   {
     auto const &function_symbol = symbol_table.lookup_ref(function_name);
     // NOLINTNEXTLINE
-    auto add_instruction = [&]() {
-      auto instruction = function.body.add_instruction();
+    auto add_instruction = [&](goto_programt::instructiont &&i) {
+      auto instruction = function.body.add(std::move(i));
       instruction->source_location = function_symbol.location;
       return instruction;
     };
```
Comment (tautschnig) [inline src/goto-instrument/generate_function_bodies.cpp]: Does this lambda still have any value left in it? The `make_*` calls would now also take a source location.

File: src/goto-instrument/generate_function_bodies.cpp [hunk 5]
```diff
@@ -96,22 +94,21 @@ class assert_false_generate_function_bodiest : public generate_function_bodiest
...
     auto const &function_symbol = symbol_table.lookup_ref(function_name);
     // NOLINTNEXTLINE
-    auto add_instruction = [&]() {
-      auto instruction = function.body.add_instruction();
+    auto add_instruction = [&](goto_programt::instructiont &&i) {
+      auto instruction = function.body.add(std::move(i));
       instruction->source_location = function_symbol.location;
       instruction->source_location.set_function(function_name);
       return instruction;
     };
```
Comment (tautschnig) [inline src/goto-instrument/generate_function_bodies.cpp]: As above: I'm not sure this Lambda adds value. Though there also is #2149...

File: src/goto-instrument/generate_function_bodies.cpp [hunk 6]
```diff
@@ -126,25 +123,22 @@ class assert_false_then_assume_false_generate_function_bodiest
...
```
Comment (tautschnig) [inline src/goto-instrument/generate_function_bodies.cpp]: As above: the Lambda...

File: src/goto-instrument/generate_function_bodies.cpp [hunk 7]
```diff
@@ -209,8 +203,8 @@ class havoc_generate_function_bodiest : public generate_function_bodiest
   {
     auto const &function_symbol = symbol_table.lookup_ref(function_name);
     // NOLINTNEXTLINE
-    auto add_instruction = [&]() {
-      auto instruction = function.body.add_instruction();
+    auto add_instruction = [&](goto_programt::instructiont &&i) {
+      auto instruction = function.body.add(std::move(i));
       instruction->source_location = function_symbol.location;
       return instruction;
     };
```
Comment (tautschnig) [inline src/goto-instrument/generate_function_bodies.cpp]: Another Lambda...

File: src/goto-programs/goto_program.h [hunk 8]
```diff
@@ -389,6 +389,7 @@ class goto_programt
     void complete_goto(targett _target)
     {
       PRECONDITION(type == INCOMPLETE_GOTO);
+      code.make_nil();
```
Comment (tautschnig) [inline src/goto-programs/goto_program.h]: Why is this necessary?

File: unit/goto-programs/goto_program_assume.cpp [hunk 9]
```diff
@@ -38,7 +38,7 @@ SCENARIO(
     symbol_table.insert(symbol);
     symbol_table.insert(fun_symbol);
     namespacet ns(symbol_table);
-    instructions.back().make_assertion(x_le_10);
+    instructions.back() = goto_programt::make_assertion(x_le_10);
```
Comment (tautschnig) [inline unit/goto-programs/goto_program_assume.cpp]: Can we actually remove the `emplace_back` above and use `add` instead?

File: unit/goto-programs/goto_program_dead.cpp [hunk 10]
```diff
@@ -33,10 +33,7 @@ SCENARIO(
 
     goto_functiont goto_function;
     auto &instructions = goto_function.body.instructions;
-    instructions.emplace_back(goto_program_instruction_typet::DEAD);
-    code_deadt removal(var_a);
-    instructions.back().make_dead();
-    instructions.back().code = removal;
+    instructions.emplace_back(goto_programt::make_dead(var_a));
```
Comment (tautschnig) [inline unit/goto-programs/goto_program_dead.cpp]: Could we use `add` instead so as to avoid use of an implementation detail?

File: unit/goto-programs/goto_program_declaration.cpp [hunk 11]
```diff
@@ -33,9 +33,7 @@ SCENARIO(
 
     goto_functiont goto_function;
     auto &instructions = goto_function.body.instructions;
-    instructions.emplace_back(goto_program_instruction_typet::DECL);
-    code_declt declaration(var_a);
-    instructions.back().make_decl(declaration);
+    instructions.emplace_back(goto_programt::make_decl(var_a));
```
Comment (tautschnig) [inline unit/goto-programs/goto_program_declaration.cpp]: As above: use `add`

File: unit/goto-programs/goto_program_goto_target.cpp [hunk 12]
```diff
@@ -33,11 +33,10 @@ SCENARIO(
 
     goto_functiont goto_function;
     auto &instructions = goto_function.body.instructions;
-    instructions.emplace_back(goto_program_instruction_typet::ASSERT);
-    instructions.back().make_assertion(x_le_10);
+    instructions.emplace_back(goto_programt::make_assertion(x_le_10));
```
Comment (tautschnig) [inline unit/goto-programs/goto_program_goto_target.cpp]: Use `add`

File: unit/goto-programs/goto_program_symbol_type_table_consistency.cpp [hunk 13]
```diff
@@ -34,8 +34,7 @@ SCENARIO(
 
     goto_functiont goto_function;
     auto &instructions = goto_function.body.instructions;
-    instructions.emplace_back(goto_program_instruction_typet::ASSERT);
-    instructions.back().make_assertion(x_le_10);
+    instructions.emplace_back(goto_programt::make_assertion(x_le_10));
```
Comment (tautschnig) [inline unit/goto-programs/goto_program_symbol_type_table_consistency.cpp]: Use `add`

File: unit/goto-programs/goto_program_table_consistency.cpp [hunk 14]
```diff
@@ -33,8 +33,7 @@ SCENARIO(
 
     goto_functiont goto_function;
     auto &instructions = goto_function.body.instructions;
-    instructions.emplace_back(goto_program_instruction_typet::ASSERT);
-    instructions.back().make_assertion(x_le_10);
+    instructions.emplace_back(goto_programt::make_assertion(x_le_10));
```
Comment (tautschnig) [inline unit/goto-programs/goto_program_table_consistency.cpp]: Use `add`

**Review by tautschnig — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 02f68bb98ccd8ae2aa40cfddaff90c87069b9fc2).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100783282

File: src/goto-instrument/accelerate/disjunctive_polynomial_acceleration.cpp (hunk 2 above)
Comment (kroening) [reply src/goto-instrument/accelerate/disjunctive_polynomial_acceleration.cpp]: done

File: src/goto-instrument/accelerate/sat_path_enumerator.cpp (hunk 3 above)
Comment (kroening) [reply src/goto-instrument/accelerate/sat_path_enumerator.cpp]: done

File: jbmc/src/java_bytecode/remove_instanceof.cpp (hunk 1 above)
Comment (kroening) [reply jbmc/src/java_bytecode/remove_instanceof.cpp]: It's actually a new instruction; now passing directly to `insert_before_swap`.

File: unit/goto-programs/goto_program_assume.cpp (hunk 9 above)
Comment (kroening) [reply unit/goto-programs/goto_program_assume.cpp]: sure

File: unit/goto-programs/goto_program_declaration.cpp (hunk 11 above)
Comment (kroening) [reply unit/goto-programs/goto_program_declaration.cpp]: done

File: unit/goto-programs/goto_program_dead.cpp (hunk 10 above)
Comment (kroening) [reply unit/goto-programs/goto_program_dead.cpp]: done

**Comment by kroening** [comment]
@tautschnig The lambdas can definitively go away once we're agreed that .function isn't abused any more.

**Comment by kroening** [comment]
(Moving the actual code changes recommended here into a new PR).

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 09fde0d18b2f8138aa4087f7338b3eb58c90fb85).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100823060
